Add type hints to functions.py

tpoterba opened the pull request.

tpoterba commented:
Unfortunately, Sphinx generates weird docs for these:
![image](https://user-images.githubusercontent.com/10562794/37361747-01a484e6-26ca-11e8-95f8-65f2e3b24cd0.png)

I still think it's worth it for IDE support.

patrick-schultz commented:
Wow, hopefully there is some way (now or in the near future) to ask Sphinx to omit the type paths in the type annotations. But we can always just turn off signatures completely if it's too noisy.

patrick-schultz commented on python/hail/expr/functions.py:
```diff
@@ -43,8 +47,9 @@ def null(t):
+
 @typecheck(x=anytype, dtype=nullable(hail_type))
-def literal(x, dtype=None):
+def literal(x: Any, dtype: Optional[Union[HailType, str]] = None) -> Expression:
```
I think I prefer the equivalent `Union[HailType, str, None]`, but not strongly. Up to you.

tpoterba replied:
I prefer Optional, because it seems to better represent the branch logic -- either it's defined or it's not, and if it's defined, then either it's a str or a Type. I don't feel strongly, though.

tpoterba replied:
Ah, another thing -- this would also let us abstract it out later more easily by defining HT as `Union[HailType, str]` once.

patrick-schultz replied:
Yeah, let's go with yours.

patrick-schultz replied:
Good point.

patrick-schultz commented on python/hail/expr/functions.py:
```diff
@@ -184,7 +189,7 @@ def cond(condition, consequent, alternate):
 
 
-def case():
+def case() -> 'hail.expr.utils.CaseBuilder':
```
Why is the return annotation a string? I haven't seen that before.

tpoterba replied:
There's a circular import if I don't. Strings in type hints are used for forward references, and are evaluated after everything else has run. This is how you can have matrixtables return tables and tables return matrix tables.

patrick-schultz replied:
Cool, thanks.

patrick-schultz commented on python/hail/expr/functions.py:
```diff
@@ -2626,7 +2660,7 @@ def abs(x):
 
 @typecheck(x=oneof(expr_numeric, expr_array(expr_numeric)))
-def signum(x):
+def signum(x: Num_T) -> Num_T:
```
I would have thought this always returns `Int32`.

tpoterba replied:
you're totally right, judging from the code.

patrick-schultz commented:
These were super minor comments, feel free to dismiss.

tpoterba commented:
I addressed your comments, but I'm still not sure what to do about the horrible signatures.

I struggle to do anything in Sphinx, and haven't had any luck finding ways to turn off the type annotations in signatures so far.

patrick-schultz commented:
Have you looked at https://github.com/agronholm/sphinx-autodoc-typehints? It looks like it could potentially help. Anyways, I agree that the type hints are worth the documentation weirdness, and we'll probably figure out how to get around the doc weirdness in the future.

tpoterba commented:
I have looked at https://github.com/agronholm/sphinx-autodoc-typehints, yes. One of the major problems with it right now is that it doesn't work with forward references, like this in matrixtable:
```python
    @typecheck_method(exprs=oneof(str, Expression),
                      named_exprs=expr_any)
    def group_rows_by(self,
                      *exprs: Tuple[Union[Expression, str]],
                      **named_exprs: NamedExprs) -> 'GroupedMatrixTable':
```

The reason it doesn't work is the decorator. The forward reference strings are evaluated in the module of the function after the module is fully imported, and in this case the module is the module of the _decorator_, not _group_rows_by_. So GroupedMatrixTable is a name error.

There are a few solutions:
 - Don't use decorators anywhere. This requires a lot of work to fully port over typechecking to typecheck2, which I'm now not totally sure is even the right thing.
- A bit hacky: import `hail` in the typecheck module and use fully clarified paths in forward references.
- see if we can eagerly evaluate the hints in the typecheck module and set them on the decorated function.

patrick-schultz commented:
Is the forward reference issue a problem with sphinx-autodoc-typehints, or with type hints in general (I don't see the connection with docs)? If sphinx-autodoc-typehints mostly works except for forward references, it looks like the return type hint is ignored if an `:rtype:` is given manually in the docstring, so we could just override it in those few cases.

tpoterba commented:
No, this is a problem with type hints and decorators in general:
```
In [6]: get_type_hints(hl.GroupedMatrixTable.group_rows_by)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-6-37e786808f16> in <module>()
----> 1 get_type_hints(hl.GroupedMatrixTable.group_rows_by)

~/anaconda2/envs/py3/lib/python3.6/typing.py in get_type_hints(obj, globalns, localns)
   1537         if isinstance(value, str):
   1538             value = _ForwardRef(value)
-> 1539         value = _eval_type(value, globalns, localns)
   1540         if name in defaults and defaults[name] is None:
   1541             value = Optional[value]

~/anaconda2/envs/py3/lib/python3.6/typing.py in _eval_type(t, globalns, localns)
    348 def _eval_type(t, globalns, localns):
    349     if isinstance(t, TypingMeta) or isinstance(t, _TypingBase):
--> 350         return t._eval_type(globalns, localns)
    351     return t
    352

~/anaconda2/envs/py3/lib/python3.6/typing.py in _eval_type(self, globalns, localns)
    243                 localns = globalns
    244             self.__forward_value__ = _type_check(
--> 245                 eval(self.__forward_code__, globalns, localns),
    246                 "Forward references must evaluate to types.")
    247             self.__forward_evaluated__ = True

<string> in <module>()

NameError: name 'GroupedMatrixTable' is not defined
```

tpoterba commented:
I don't think that ignoring is good enough -- this will throw an exception which I don't think the extension will handle. This is what happened when I just tried to throw the extension into the mix and build the docs (and it happened on a function call that included an explicit return type in the docstring)

patrick-schultz commented:
Ok, I think I get it now. The hacky solution (import `hail` in `typecheck`) doesn't seem too bad, especially if we document it so we remember to undo it if we find a better solution.

tpoterba commented:
Yeah, good call. 

If we can decorate, then we can do implicit argument conversions...

Let's talk this afternoon.
